compute_tools/src/monitor.rs

```diff
@@ -6,197 +6,296 @@ use chrono::{DateTime, Utc};
 use compute_api::responses::ComputeStatus;
 use compute_api::spec::ComputeFeature;
 use postgres::{Client, NoTls};
-use tracing::{debug, error, info, warn};
+use tracing::{Level, error, info, instrument, span};
 
 use crate::compute::ComputeNode;
+use crate::metrics::{PG_CURR_DOWNTIME_MS, PG_TOTAL_DOWNTIME_MS};
 
 const MONITOR_CHECK_INTERVAL: Duration = Duration::from_millis(500);
 
-// Spin in a loop and figure out the last activity time in the Postgres.
-// Then update it in the shared state. This function never errors out.
-// NB: the only expected panic is at `Mutex` unwrap(), all other errors
-// should be handled gracefully.
-fn watch_compute_activity(compute: &ComputeNode) {
-    // Suppose that `connstr` doesn't change
-    let connstr = compute.params.connstr.clone();
-    let conf = compute.get_conn_conf(Some("compute_ctl:activity_monitor"));
-
-    // During startup and configuration we connect to every Postgres database,
-    // but we don't want to count this as some user activity. So wait until
-    // the compute fully started before monitoring activity.
-    wait_for_postgres_start(compute);
-
-    // Define `client` outside of the loop to reuse existing connection if it's active.
-    let mut client = conf.connect(NoTls);
-
-    let mut sleep = false;
-    let mut prev_active_time: Option<f64> = None;
-    let mut prev_sessions: Option<i64> = None;
-
-    if compute.has_feature(ComputeFeature::ActivityMonitorExperimental) {
-        info!("starting experimental activity monitor for {}", connstr);
-    } else {
-        info!("starting activity monitor for {}", connstr);
+struct ComputeMonitor {
+    compute: Arc<ComputeNode>,
+
+    // The moment when Postgres had some activity,
+    // that should prevent compute from being suspended.
+    last_active: Option<DateTime<Utc>>,
+
+    // The moment when we last tried to check Postgres.
+    last_checked: DateTime<Utc>,
+    // The last moment we did a successful Postgres check.
+    last_up: DateTime<Utc>,
+
+    // These two attributes are purely for statistics change tracking,
+    // they can be outdated.
+    active_time: Option<f64>,
+    sessions: Option<i64>,
+
+    // Use experimental statistics-based activity monitor. It's no longer
+    // 'experimental' per se, as it's enabled for everyone, but we still
+    // keep the flag as an option to turn it off in some cases if it will
+    // misbehave.
+    experimental: bool,
+}
+
+impl ComputeMonitor {
+    fn report_down(&mut self) {
```

Comment:
It isn't clear to me why this is a mutable reference.

Reply:
Probably left from the time when I was doing some changes in it, not only emitting a metric. Thanks